File: 5. LinkedLists/get_node_value/get_node_value.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");

// a.next = b;
// b.next = c;
// c.next = d;

// Iterative
const getNodeValue = (head, index) => {
    while(head !== null){
        if(index-- === 0) return head.val;
        head = head.next;
    }
    return null;
}

// Recursive
// const getNodeValue = (head, index) => {
//     if(head === null) return null;
//     if(index === 0) return head.val;
//     return getNodeValue(head.next, index-1);
// }

// Complexity
// Iterative
// n = number of nodes
// Time: O(n)
// Space: O(1)

// Recursive
// n = number of nodes
// Time: O(n)
// Space: O(n)


// let ans = getNodeValue(a, 2);
// console.log(`Ans: ${ans}`);

File: 5. LinkedLists/4. get_node_value/get_node_value.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d

# def get_node_value(head, index):
#   current = head
#   while(current is not None):
#     if(index == 0):
#       return current.val
#     index -= 1
#     current = current.next
  
def get_node_value(head, index):
  if index == 0:
    return head.val
  if head is None:
    return None
  index -= 1
  return get_node_value(head.next, index)
  
print(f"Answer: {get_node_value(a, 2)}") # 'c'